### ISSUE TITLE
java.lang.IllegalArgumentException: Image byte buffer must be allocated as 'direct'

### ISSUE THREAD

**Issue by alexanderyellow** [googlesamples/android-vision#33]
Hi, I was testing the sample in my Nexus 5. When i press "Read Barcode" button i get following exception:
java.lang.IllegalArgumentException: Image byte buffer must be allocated as 'direct'. See ByteBuffer.allocateDirect().
            at com.google.android.gms.vision.Frame$Builder.setImageData(Unknown Source)
            at barcodereader.camera.CameraSource$FrameProcessingRunnable.run(CameraSource.java:930)

in this snippet:
outputFrame = new Frame.Builder()
                            .setImageData(mPendingFrameData, mPreviewSize.getWidth(),
                                    mPreviewSize.getHeight(), ImageFormat.NV21)
                            .setId(mPendingFrameId)
                            .setTimestampMillis(mPendingTimeMillis)
                            .setRotation(mRotation)
                            .build();

Please, help me resolve this problem!

**Comment by alexanderyellow** [comment]
I've solved this problem:
setImageData(ByteBuffer.allocateDirect(mPendingFrameData.capacity()), mPreviewSize.getWidth(),
                                    mPreviewSize.getHeight(), ImageFormat.NV21);
I think it's correct solution.

**Comment by pm0733464** [comment]
That demo requires compiling with Google Play Services 8.1.  I'm guessing that you compiled with Google Play Services 7.8.

This was the result of a bug fix that occurred between the two versions.  The issue is that using direct byte buffers with the CameraSource class will result in a pixel shift of the images sent to the detector (e.g., the position reported for the barcode may be shifted to the left several pixels).  In 8.1 we added support for indirect byte buffers to address this, but the 7.8 library doesn't have that ability.

Long story short:  Compile with 8.1 if you can.  If you can't, you'll need to change to direct byte buffers as suggested above, and positions reported back will be off by several pixels.

**Comment by alexanderyellow** [comment]
Thanks alot for your fast and useful answer. It's helped me.

**Comment by cruddell001** [comment]
It would be helpful if the Barcode codelabs course that references this sample project were updated to reflect the correct version.  https://codelabs.developers.google.com/codelabs/bar-codes/#2
